### ISSUE TITLE
val_dataloader shuffle True/False yields differnt results

### ISSUE THREAD

**Issue by sirtris** [PyTorchLightning/pytorch-lightning#9017]
## 🐛 Bug

Setting the `shuffle` argument in the validation `DataLoader` in the DataModule to `True` or `False` results in different (in my case dice) scores. This is for the train dice as well as for the validation dice. 

### Expected behavior

As far as I understood from the pytroch forum this problem should be solved by calling `model.eval()` before running the validation batches. This should of course be done by the `validation_step` method in the LightingModel.  
I also don't understand how shuffling of the validation set would influence the train performance.  
I am using `torchmetrics.functional.dice_score` to compute the dice scores.

### Environment

- PyTorch Lightning Version: 1.4.2
- PyTorch Version: 1.8.1
- Python version: 3.8.10
- OS: Pop!_OS 20.10
- CUDA/cuDNN version: 11.2

**Comment by rohitgr7** [comment]
`model.eval()` is set by default by lightning before validation starts, so that shouldn't be a problem here. Setting `shuffle=True` in train_dataloader might lead to different results since the loss at each iteration will vary. In the case of valid_dataloader ideally shuffle=True/False shouldn't vary but here your metric is dice score and dice score doesn't just do the average of the individual dice scores for each image in that batch. Dice score for a batch varies a little and is not equal to the average score for each image. So when you are assigning `shuffle=True` in valid_dataloader, the batch in each iteration varies, and therefore your dice score too.

**Comment by sirtris** [comment]
Thanks for the clarification about the dice score. 
What still is unclear to me how shuffle in the validation dataloader can affect the dice during training. 
Btw. I used `pl.seed_everything` and the deterministic flag in the Trainer.

**Comment by rohitgr7** [comment]
It should not if you have `shuffle=False` in train_dataloader.

**Comment by tchaton** [comment]
Dear @sirtris,

I believe @rohitgr7 provided you with the right answer and therefore I will be closing this issue.
Feel free to re-open that issue if you have more questions.

Best,
T.C

**Comment by sirtris** [comment]
Hi,
Unfortunately I don't think the issue is solved for me. 
I still don't understand why setting shuffle to True/False gives different dice scores for the train batches. They should not be influenced by things happening during the validation step right?

I am using `torchmetrics.functional.dice_score`. When I would use the module API I would make one object for the train dice and a separate one for the validation dice. With the functional API I cannot do this. Might this have something to do with the problem?

**Comment by rohitgr7** [comment]
functional metrics doesn't store any state so validation batches have no influence on train batches. Is your training loss also different when you set shuffle=True/False in val_dataloader?? Some schedulers like ReduceLROnPlateau can influence this thing in case you are using it and monitoring some validation metric to update the LR of the optimizer.

**Comment by sirtris** [comment]
The train loss is almost the same. There is a very slight difference but only in the third decimal place. 
The difference between the dice score is 0.4855 (shuffle=true) vs 0.3112 (shuffle=False), for the most extreme epoch.

I use `CosineAnnealingLR` as Lr scheduler, so I think that should not have any effects here.

**Comment by sirtris** [comment]
I tried using `check_val_every_n_epoch` to remove the validation step altogether and got yet another train dice.

**Comment by rohitgr7** [comment]
pretty weird :sweat_smile: training is completely independent from validation. Is it possible to get a minimal reproducible example? I can try and check the issue.

**Comment by sirtris** [comment]
I used the code from [Use Pytorch Lightning with Weights & Biases](https://wandb.ai/cayush/pytorchlightning/reports/Use-Pytorch-Lightning-with-Weights-Biases--Vmlldzo2NjQ1Mw) to create a [minimal reproducible example](https://colab.research.google.com/drive/1IjccvDjD6XzRHR8ud8eyHSEvytfaj3VW?usp=sharing).  

Obviously it uses wandb to log the results. On my local machine I could also recreate the issue using tensor board. (No idea how to use it on colab).  
 
All I changed was:  
- seeded everything and use the `deterministic` flag in the Trainer.
- used `torchmetrics`  to compute the accuracy
- `shuffle=True/False` in the validation loader

The rest is as the original example.

**Comment by rohitgr7** [comment]
Need access to run the notebook. @sirtris

**Comment by sirtris** [comment]
Sorry, my bad.
Should work now

**Comment by rohitgr7** [comment]
I ran the notebook, by setting shuffle=False/True separately in different runs, and it seems to be generating the same metrics as expected. Did you reset the seed when you toggle the shuffle in val_dataloader?
